src/algebra/jordan/special.lean

```diff
@@ -0,0 +1,191 @@
+
+instance [has_zero α] : has_zero (αˢʸᵐ) := { zero := sym 0 }
+instance [has_sub α] : has_sub αˢʸᵐ := { sub := λ a b, sym (unsym a - unsym b) }
+
+instance [has_add α] : has_add αˢʸᵐ :=
+{ add := λ a b, sym (unsym a + unsym b) }
+
+@[simp] lemma sym_add [has_add α] (a b : α) : sym (a + b) = sym a + sym b := rfl
+@[simp] lemma unsym_add [has_add α] (x y : αˢʸᵐ) : unsym (x + y) = unsym x + unsym y := rfl
+
+
+instance [has_neg α] : has_neg αˢʸᵐ :=
+{ neg := λ a, sym ( -(unsym a)) }
+
+instance {R : Type*} [has_scalar R α] : has_scalar R αˢʸᵐ :=
+{ smul := λ r a, sym (r•(unsym a)) }
+
+@[simp] lemma sym_smul {R : Type*} [has_scalar R α] (c : R) (a : α) : sym (c • a) = c • sym a := rfl
+
+lemma unsym_injective : function.injective (@unsym α) := λ _ _, id
+
+instance [add_comm_semigroup α] : add_comm_semigroup (αˢʸᵐ) :=
+unsym_injective.add_comm_semigroup _ (λ _ _, rfl)
+
+instance [add_monoid α] : add_monoid (αˢʸᵐ) :=
+unsym_injective.add_monoid_smul _ rfl (λ _ _, rfl) (λ _ _, rfl)
+
+instance [add_group α] : add_group (αˢʸᵐ) :=
+unsym_injective.add_group_smul _ rfl
+  (λ _ _, rfl) (λ _, rfl) (λ _ _, rfl) (λ _ _, rfl) (λ _ _, rfl)
+
+instance [add_comm_monoid α] : add_comm_monoid (αˢʸᵐ) :=
+{ ..sym_alg.add_comm_semigroup, ..sym_alg.add_monoid }
+
+instance [add_comm_group α] : add_comm_group (αˢʸᵐ) :=
+{ ..sym_alg.add_comm_monoid, ..sym_alg.add_group }
+
+
+
+@[simp] lemma sym_zero [has_zero α] : sym (0 : α) = 0 := rfl
+@[simp] lemma unsym_zero [has_zero α] : unsym (0 : αˢʸᵐ) = 0 := rfl
+
+@[simp] lemma unsym_eq_zero_iff {α} [has_zero α] (a : αˢʸᵐ) : a.unsym = (0 : α) ↔ a = (0 : αˢʸᵐ) :=
+unsym_injective.eq_iff' rfl
+
+lemma zero_zero [ring α] : unsym add_comm_group.zero = (0:α) := rfl
+
+instance {R : Type*} [semiring R] [add_comm_monoid α] [module R α] : module R αˢʸᵐ :=
+function.injective.module R ⟨unsym, rfl, λ _ _, rfl⟩ (λ _ _, id) (λ _ _, rfl)
+
+/- Introduce the symmetrised multiplication-/
+instance [has_add α] [has_mul α] [has_one α] [invertible (2 : α)] : has_mul(αˢʸᵐ) :=
+{ mul := λ a b, sym ((⅟2)*(unsym(a)*unsym(b) + unsym(b)*unsym(a))) }
+
+lemma mul_def [ring α] [invertible (2 : α)] (a b: αˢʸᵐ) :
+  a*b = sym ((⅟2)*(unsym(a)*unsym(b)+unsym(b)*unsym(a))) := by refl
+
+/- The symmetrisation of a real (unital, associative) algebra is a non-associative ring -/
+instance [ring α] [invertible (2 : α)] : non_unital_non_assoc_ring (αˢʸᵐ) :=
+{ zero_mul := λ _,
+  begin
+    simp only [mul_def,zero_zero, add_zero, sym_zero, zero_mul, mul_zero, smul_zero],
+    exact rfl,
+  end,
+  mul_zero :=  λ _,
+  begin
+    simp only [mul_def,zero_zero, add_zero, sym_zero, zero_mul, mul_zero, smul_zero],
+    exact rfl,
+  end,
+  left_distrib := λ a b c, begin
+    change (⅟2)*(unsym(a)*(unsym(b)+unsym(c))+(unsym(b)+unsym(c))*unsym(a)) =
+      (⅟2)*(unsym(a)*unsym(b)+unsym(b)*unsym(a))+(⅟2)*(unsym(a)*unsym(c)+unsym(c)*unsym(a)),
+    rw [←mul_add, mul_add (unsym a), add_mul, ← add_assoc, ← add_assoc],
+    finish,
+  end,
```

Comment:
If you had functions like
```lean
def sym.map (f : α → β) : αˢʸᵐ → βˢʸᵐ := λ a, sym (f (unsym a))
def sym.map₂ (f : α → β → γ) : αˢʸᵐ → βˢʸᵐ → γˢʸᵐ := λ a b, sym (f (unsym a) (unsym b))
def sym.map₃ (f : α → β → γ → δ) : αˢʸᵐ → βˢʸᵐ → γˢʸᵐ → δˢʸᵐ :=
λ a b c, sym (f (unsym a) (unsym b) (unsym c))
```
(untested) you could simplify code like this to (also untested)
```lean
left_distrib := sym.map₃ $ λ a b c, begin
    change (⅟2)*(a*(b+c)+(b+c)*a) = (⅟2)*(a*b+b*a)+(⅟2)*(a*c+c*a),
    rw [←mul_add, mul_add a, add_mul, ← add_assoc, ← add_assoc],
    finish,
  end,
```
By the way, many of the parentheses in the original code are unnecessary. Lean parses `unsym(a)*(unsym(b)+unsym(c))` and `unsym a * (unsym b + unsym c)` just the same (and the second would be more common in mathlib).

There's also been a project to eliminate `finish` from mathlib since there's not a plan to port it to Lean 4.

Reply:
@kmill Thanks. I've eliminated `finish` from the PR. I'm afraid I couldn't see how to make your other suggestion work.

Reply:
I couldn't get `sym.map₃ ` to work, so I've removed the extraneous brackets instead.

Reply:
You want `rec` not map, like `mul_opposite.rec` etc.

Reply:
Here's a shorter proof:
```lean
  left_distrib := λ a b c, match a, b, c with
    | sym a, sym b, sym c := begin
      rw [sym_mul_sym, sym_mul_sym, ←sym_add, sym_mul_sym, ←sym_add, mul_add a, add_mul _ _ a,
        add_add_add_comm, mul_add],
    end
  end,
```
which needs `@[pattern]` on `sym`.

Reply:
This is what ended up getting merged (in a different PR) https://github.com/leanprover-community/mathlib/blob/90e2957ebf4bd208c27795dd16ced767f8466270/src/algebra/symmetrized.lean#L181